Draco compression support for GLB files

mvaligursky opened the pull request:
Fixes #2001

Draco mesh compression is now supported within GLB files. See library here: https://github.com/google/draco

Draco modules version: 1.3.6

Test stats on Bed Model:

<img width="580" alt="Screen Shot 2020-04-23 at 11 58 11 AM" src="https://user-images.githubusercontent.com/59932779/80094362-05443e80-855e-11ea-8cf0-ede14394517e.png">

- vertices: 496495
- triangles: 801397
- uncompressed GLB: 22.8MB (95ms to extract meshes on Macbook Pro)
- compressed GLB: 2MB (990ms to decompress meshes on Macbook Pro)

willeastcott commented:
I'd suggest placing the draco WASM module in `examples\assets\lib\draco`. You could also rename the files with a similar naming convention. e.g.:

```
draco.js
draco.wasm.js
draco.wasm.wasm
```

mvaligursky commented:
I've moved and renamed draco library files as suggested, and addressed lint issues.
stil WIP, looking at supporting various types instead of using float32 for all attribute arrays when decompressing

willeastcott commented:
Nearly but not quite. 😄  `draco_wasm.js` -> `draco.wasm.js`

willeastcott commented on src/resources/parser/glb-parser.js:
```diff
@@ -98,18 +98,31 @@ Object.assign(pc, function () {
+        }
+
+        // generate normals
+        var normals = Float32Array.from(pc.calculateNormals(positions, indices));
```
Sorry. This doesn't work in IE11: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/from

mvaligursky replied:
I'll replace this .. this is existing code in main, so perhaps we wait till fix is in.

mvaligursky replied:
fixed

slimbuck commented on src/resources/parser/glb-parser.js:
```diff
@@ -298,24 +388,120 @@ Object.assign(pc, function () {
+                                    status = decoder.DecodeBufferToMesh(buffer, outputGeometry);
+                                    break;
+                                default:
+                                case decoderModule.INVALID_GEOMETRY_TYPE:
```
would you mind putting ```default:``` last?

mvaligursky replied:
done

slimbuck commented on src/resources/parser/glb-parser.js:
```diff
@@ -298,24 +388,120 @@ Object.assign(pc, function () {
+                            if (!status.ok() || outputGeometry.ptr == 0) {
+                                // #ifdef DEBUG
+                                console.error("Failed to decode draco compressed asset: " + status.error_msg());
+                                // #endif
```
should the draco code below continue to be executed if this fails? shouldn't this result in load failure?

mvaligursky replied:
done

slimbuck commented on src/resources/parser/glb-parser.js:
```diff
@@ -298,24 +388,120 @@ Object.assign(pc, function () {
+                            var buffer = new decoderModule.DecoderBuffer();
+                            buffer.Init(uint8Buffer, uint8Buffer.length);
+
+                            var decoder = new decoderModule.Decoder();
```
how slow is it to construct a decoder instance? would there be any benefit in caching it?

slimbuck commented on src/resources/parser/glb-parser.js:
```diff
@@ -327,7 +513,7 @@ Object.assign(pc, function () {
                     indexFormat = pc.INDEXFORMAT_UINT32;
                 }
-                var numIndices = indices.length;
+                numIndices = indices.length;
```
can probably remove numIndices here (or make line 519 also use it).

mvaligursky replied:
done

willeastcott commented on src/resources/parser/glb-parser.js:
```diff
@@ -253,6 +207,142 @@ Object.assign(pc, function () {
+                    break;
+
+                default:
+                    console.error("Element type not implemented: " + storageType);
```
I'd wrap this (and any console logging) with an #ifdef DEBUG

mvaligursky replied:
done

marklundin commented:
@mvaligursky I didn't realise Draco decompression was already in the engine! So theoretically, if we had a local copy of a project, we could run through all the glb files, compress them with Draco and replace them, then as long as we include the Draco lib in the global object, we can just run the project and it will automatically decompress and use them as normal? 🙏

Maksims commented:
Does the GLB parsing+decompression happen in a web worker?

mvaligursky commented:
@marklundin - yes, this should work fine, I don't expect a problem with this. I think to get the loader working, you just need to drop those 3 draco libraries into Editor project, that should work.
https://github.com/playcanvas/engine/tree/master/examples/lib/draco

But as @Maksims mentioned - the decompression takes place on the main thread at the moment, so it costs a bit. See some timing in the description of this PR.

marklundin commented:
Ok that's good to know! I think in three.js they [fire up a worker](<script src='/decoder/draco.js' />, but I'd be keento know what how performant the decompression is.

Also have you tried with the latest versions? https://github.com/google/draco/tree/master/javascript

marklundin commented:
on a side note compiling the shaders in a worker would be amazing 😆

mvaligursky commented:
yeah, there's lots we want to do.
This is on the list as well .. compresses just as well as draco, but decompression is a lot faster: https://github.com/zeux/meshoptimizer
and handles sparse morph targets unlike draco.
